mlflow/projects/__init__.py

```diff
@@ -189,6 +194,31 @@ def _is_local_uri(uri):
     return not _GIT_URI_REGEX.match(uri)
 
 
+def _is_valid_version(repo, version):
+    """Verify the commit version of branch name is valid, raise exception if otherwise."""
+    from git.exc import GitCommandError
+    try:
+        return repo.git.rev_parse("--verify", version)
+    except GitCommandError:
+        # Second attempt to see if this is a branch
+        try:
+            return repo.git.rev_parse("--verify", "refs/remotes/origin/%s" % version)
+        except GitCommandError as e:
+            raise ExecutionException("Got exception when validating git version '%s' "
+                                     "- please ensure that the version exists in the repo. "
+                                     "Error: %s" % (version, e))
+
+
+def _is_valid_branch_name(uri, work_dir, branch):
+    """Returns True if the branch exists in the Git project."""
+    if branch is not None:
+        sub_directory = _parse_subdirectory(uri)[1]
+        from git import Repo
+        repo = Repo(work_dir.replace(sub_directory, ''), search_parent_directories=True)
+        return repo.git.rev_parse("--verify", "refs/heads/%s" % branch)
```

Comment:
AFAICT this will actually raise an exception if `refs/heads/<branch>` is not a valid branch right? Should we wrap it in a `try`/`except GitCommandError` block? It'd be nice to add a unit test for this method in tests/test_projects.py. We could try calling `_is_valid_branch_name` on `local_git_repo` with a valid & invalid branch name and verify the return value in each case

Reply:
Gotcha - I think actually the previous `git.checkout(version)` will verify that `version` is a valid git ref (commit hash, tag name, or branch name). Thus when we call `_is_valid_branch_name`, the `branch` argument here could actually be a commit hash or tag name, which won't exist in refs/heads & raise an exception here. You can see this experimentally by running `mlflow run git@github.com:databricks/mlflow-example.git -P alpha=0.1 --no-conda --version 7193f0c`.

Reply:
True. Thanks for catching this.